### PR TITLE
daemon,overlord: add subcommand handling to snapctl

### PR DESCRIPTION
**Pull request by kyrofa**
Make some more progress on LP: [#1586465](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/1586465) by adding some more logic and subcommands to `snapctl`. Stub out initial `set` command. This should allow other subcommands to be easily added (e.g. for interface hooks).

### CODE REVIEW

File: overlord/hookstate/handler.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+// addHandler adds a new handler to the collection. Returns the ID which can be
+// used to refer to this handler in the future.
+func (c *handlerCollection) addHandler(handler Handler) (string, error) {
+	c.handlersMutex.Lock()
+	defer c.handlersMutex.Unlock()
+
+	// Generate a secure, random ID for this handler
+	idBytes := make([]byte, 32)
+	_, err := rand.Read(idBytes)
```
Comment (zyga) [inline overlord/hookstate/handler.go]: Can this block for any amount of time?

Comment (kyrofa) [inline overlord/hookstate/handler.go]: It uses `getrandom()`, which by default uses `/dev/urandom`, which shouldn't block.

File: overlord/hookstate/handler.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+// handlerCount returns the number of handlers in the collection.
+func (c *handlerCollection) handlerCount() int {
+	c.handlersMutex.RLock()
+	defer c.handlersMutex.RUnlock()
+
+	return len(c.handlers)
+}
+
+func (c *handlerCollection) getHandler(context string) (Handler, error) {
```
Comment (mvo5) [inline overlord/hookstate/handler.go]: It looks like this one is not tested?

Comment (kyrofa) [inline overlord/hookstate/handler.go]: Oops, good catch, thanks! Fixed.

File: overlord/hookstate/hookmgr.go [hunk 3]
```diff
@@ -75,9 +66,10 @@ type hookSetup struct {
 func Manager(s *state.State) (*HookManager, error) {
 	runner := state.NewTaskRunner(s)
 	manager := &HookManager{
-		state:      s,
-		runner:     runner,
-		repository: newRepository(),
+		state:          s,
+		runner:         runner,
+		repository:     newRepository(),
+		activeHandlers: newHandlerCollection(),
```
Comment (niemeyer) [inline overlord/hookstate/hookmgr.go]: s/activeHandlers/handlers/ (we don't have any inactive ones)

I've mentioned this in earlier reviews: it doesn't feel like the additional abstraction there is gaining us much. As a thought experiment, try to inline a map and a simple mutex here, and see how much code you need to write here, vs. how much code you drop.

That said, I believe what we really want here is a map of context, not handlers. The reason being that context is a concrete and common implementation, and it would be simple and convenient to introduce a `HookHandler` method on context to navigate from a context into its handler. Doing that in the opposite direction is boring, because every single handler will need to reimplement the same method to return its context.

File: overlord/hookstate/hookmgr.go [hunk 4]
```diff
@@ -153,13 +150,19 @@ func (m *HookManager) doRunHook(task *state.Task, tomb *tomb.Tomb) error {
 
 	handler := handlers[0]
 
+	context, err := m.activeHandlers.addHandler(handler)
```
Comment (niemeyer) [inline overlord/hookstate/hookmgr.go]: There's a bit of terminology confusion here which would be good to avoid. What we have here is not the context. It's the context ID, and the actual context is in line 142 above:
```
handlers := m.repository.generateHandlers(&Context{task: task, setup: setup})
```
That value being created inline and passed as a parameter is the actual context, and we should probably get it with NewContext(task, setup) so it has a chance to generate its own ID internally. We can then:
```
m.context[context.ID()] = context
```
So, map of context, a context ID, and the context.

File: overlord/hookstate/hookmgr.go [hunk 5]
```diff
@@ -119,10 +111,15 @@ func (m *HookManager) Stop() {
 	m.runner.Stop()
 }
 
-// SnapCtl is used to conduct the action required by the snapctl request.
-func (m *HookManager) SnapCtl(request SnapCtlRequest) (stdout, stderr []byte, err error) {
-	// TODO: Make this do something useful
-	return nil, nil, nil
+// GetHandler obtains the handler for the given context.
+func (m *HookManager) GetHandler(context string) (Handler, error) {
```
Comment (niemeyer) [inline overlord/hookstate/hookmgr.go]: As a natural consequence of the other points:
```
func (m *HookManager) Context(id string) (*Context, error)
```

File: overlord/hookstate/tools/tools.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package tools contains the various snapctl subcommands.
+package tools
```
Comment (niemeyer) [inline overlord/hookstate/tools/tools.go]: Sorry, I know it was my suggestion, but can we please call this package as "ctlcmd"?

File: overlord/hookstate/tools/tools.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+	stdout  io.Writer
+	stderr  io.Writer
+	handler hookstate.Handler
+}
+
+func (c *baseCommand) setStdout(w io.Writer) {
+	c.stdout = w
+}
+
+func (c *baseCommand) writeStdout(format string, a ...interface{}) {
```
Comment (niemeyer) [inline overlord/hookstate/tools/tools.go]: In Go "write" is generally about raw byte writes. Can we please name this `printf`?

File: overlord/hookstate/tools/tools.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+	if c.stdout != nil {
+		c.stdout.Write([]byte(fmt.Sprintf(format, a...)))
+	}
+}
+
+func (c *baseCommand) setStderr(w io.Writer) {
+	c.stderr = w
+}
+
+func (c *baseCommand) writeStderr(format string, a ...interface{}) {
```
Comment (niemeyer) [inline overlord/hookstate/tools/tools.go]: Similarly, `errorf`?

File: overlord/hookstate/tools/tools.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+	if c.stderr != nil {
+		c.stderr.Write([]byte(fmt.Sprintf(format, a...)))
+	}
+}
+
+func (c *baseCommand) setHandler(handler hookstate.Handler) {
```
Comment (niemeyer) [inline overlord/hookstate/tools/tools.go]: Rather than set/getHandler, it's probably nicer to do:
```
Execute(context, args)
```
Per points above, we can get the handler out of the context when that's necessary.

File: overlord/hookstate/tools/tools.go [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+func addCommand(name string, command toolCommand) {
+	if commands == nil {
+		commands = make(map[string]toolCommand)
+	}
+	commands[name] = command
+}
+
+// RunCommand runs the requested command.
+func RunCommand(handler hookstate.Handler, args []string) (stdout, stderr []byte, err error) {
+	parser := flags.NewParser(nil, flags.PassDoubleDash|flags.PassAfterNonOption)
```
Comment (niemeyer) [inline overlord/hookstate/tools/tools.go]: Can we please try to keep PassAfterNonOption out for the time being?

File: overlord/hookstate/tools/tools.go [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+func addCommand(name string, command toolCommand) {
+	if commands == nil {
+		commands = make(map[string]toolCommand)
+	}
+	commands[name] = command
+}
+
+// RunCommand runs the requested command.
+func RunCommand(handler hookstate.Handler, args []string) (stdout, stderr []byte, err error) {
```
Comment (niemeyer) [inline overlord/hookstate/tools/tools.go]: Also related to earlier points, context instead of handler.

File: overlord/hookstate/tools/tools.go [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+	parser := flags.NewParser(nil, flags.PassDoubleDash|flags.PassAfterNonOption)
+
+	// Create stdout/stderr buffers, and make sure commands use them.
+	var stdoutBuffer bytes.Buffer
+	var stderrBuffer bytes.Buffer
+	for name, command := range commands {
+		command.setStdout(&stdoutBuffer)
```
Comment (niemeyer) [inline overlord/hookstate/tools/tools.go]: Something isn't right here. This is mutating the global map of commands.

File: overlord/hookstate/tools/tools.go (hunk 9 above)
Comment (kyrofa) [inline overlord/hookstate/tools/tools.go]: If I use go-flags I have to conform to the Commander interface, which enforces that `Execute(args)` signature. Otherwise I'd completely agree.

Would you prefer that this function is `setContext()` instead, however?

File: overlord/hookstate/hookmgr.go (hunk 3 above)
Comment (kyrofa) [inline overlord/hookstate/hookmgr.go]: Done.

File: overlord/hookstate/hookmgr.go (hunk 4 above)
Comment (kyrofa) [inline overlord/hookstate/hookmgr.go]: Done.

File: overlord/hookstate/hookmgr.go (hunk 5 above)
Comment (kyrofa) [inline overlord/hookstate/hookmgr.go]: Done.

File: overlord/hookstate/tools/tools.go (hunk 6 above)
Comment (kyrofa) [inline overlord/hookstate/tools/tools.go]: Done.

File: overlord/hookstate/tools/tools.go (hunk 7 above)
Comment (kyrofa) [inline overlord/hookstate/tools/tools.go]: Done.

File: overlord/hookstate/tools/tools.go (hunk 8 above)
Comment (kyrofa) [inline overlord/hookstate/tools/tools.go]: Done.

File: overlord/hookstate/tools/tools.go (hunk 9 above)
Comment (kyrofa) [inline overlord/hookstate/tools/tools.go]: Implemented `setContext()`.

File: overlord/hookstate/tools/tools.go (hunk 10 above)
Comment (kyrofa) [inline overlord/hookstate/tools/tools.go]: Done.

File: overlord/hookstate/tools/tools.go (hunk 11 above)
Comment (kyrofa) [inline overlord/hookstate/tools/tools.go]: Done.

File: overlord/hookstate/tools/tools.go (hunk 12 above)
Comment (kyrofa) [inline overlord/hookstate/tools/tools.go]: Fixed via generator functions.

File: daemon/api.go [hunk 13]
```diff
@@ -1595,15 +1596,22 @@ func runSnapctl(c *Command, r *http.Request, user *auth.UserState) Response {
...
 
-	stdout, stderr, err := c.d.overlord.HookManager().SnapCtl(snapctlRequest)
+	// Right now snapctl is only used for hooks. If at some point it grows
+	// beyond that, this probably shouldn't go straight to the HookManager.
+	context, err := c.d.overlord.HookManager().Context(snapctlRequest.ContextID)
+	if err != nil {
+		return BadRequest("cannot run snapctl: %s", err)
+	}
+
+	stdout, stderr, err := ctlcmd.RunCommand(context, snapctlRequest.Args)
```
Comment (niemeyer) [inline daemon/api.go]: This one might be just `ctlcmd.Run` now that we have cmd in the package name.

File: overlord/hookstate/context.go [hunk 14]
```diff
@@ -29,8 +31,27 @@ import (
...
-	task  *state.Task
-	setup hookSetup
+	task    *state.Task
+	setup   HookSetup
+	id      string
+	handler Handler
+}
+
+// NewContext returns a new Context.
+func NewContext(task *state.Task, setup HookSetup, handler Handler) (*Context, error) {
```
Comment (niemeyer) [inline overlord/hookstate/context.go]: s/HookSetup/*HookSetup/ (pointer)

File: overlord/hookstate/context.go [hunk 15]
```diff
@@ -29,8 +31,27 @@ import (
 
 // Context represents the context under which a given hook is running.
 type Context struct {
-	task  *state.Task
-	setup hookSetup
+	task    *state.Task
+	setup   HookSetup
```
Comment (niemeyer) [inline overlord/hookstate/context.go]: s/HookSetup/*HookSetup/ (pointer)

File: overlord/hookstate/context_test.go [hunk 16]
```diff
@@ -29,7 +29,7 @@ import (
 type contextSuite struct {
 	context *Context
 	task    *state.Task
-	setup   hookSetup
+	setup   HookSetup
```
Comment (niemeyer) [inline overlord/hookstate/context_test.go]: s/HookSetup/*HookSetup/ (pointer)

File: overlord/hookstate/ctlcmd/ctlcmd.go [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+	setStderr(w io.Writer)
+	setContext(context *hookstate.Context)
+	getContext() *hookstate.Context
+
+	Execute(args []string) error
+}
+
+type commandGenerator func() ctlCommand
+
+var commandGenerators map[string]commandGenerator
```
Comment (niemeyer) [inline overlord/hookstate/ctlcmd/ctlcmd.go]: Compound suggestion:

`var commandGenerators = make(map[string]func() command)`

 1. `ctlCommand` => `command` and use `cmd` as the variable name
 2. drop `commandGenerator` and use `func() command`, shorter and more clear
 3. inline the map creation at declaration time, so no need to test for `nil` below

File: overlord/hookstate/ctlcmd/ctlcmd.go [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+	if c.stderr != nil {
+		c.stderr.Write([]byte(fmt.Sprintf(format, a...)))
+	}
+}
+
+func (c *baseCommand) setContext(context *hookstate.Context) {
+	c.context = context
+}
+
+func (c *baseCommand) getContext() *hookstate.Context {
```
Comment (niemeyer) [inline overlord/hookstate/ctlcmd/ctlcmd.go]: Can we please name the field as `c` simply and name this as `context()` so that inside the implementations we can refer to it as the more conventional `c.context()`.

File: client/client_test.go [hunk 19]
```diff
@@ -224,12 +225,12 @@ func (cs *clientSuite) TestSnapClientIntegration(c *check.C) {
...
-	options := client.SnapCtlOptions{
-		Context: "foo",
-		Args:    []string{"bar", "--baz"},
+	request := hookstate.SnapCtlRequest{
+		ContextID: "foo",
+		Args:      []string{"bar", "--baz"},
 	}
 
-	stdout, stderr, err := cli.RunSnapctl(options)
+	stdout, stderr, err := cli.RunSnapctl(request)
```
Comment (niemeyer) [inline client/client_test.go]: This should also be received as a pointer, as the other option values in client.

File: client/snapctl.go [hunk 20]
```diff
@@ -23,25 +23,17 @@ import (
...
+)
 
 type snapctlOutput struct {
 	Stdout string `json:"stdout"`
 	Stderr string `json:"stderr"`
 }
 
 // RunSnapctl requests a snapctl run for the given options.
-func (client *Client) RunSnapctl(options SnapCtlOptions) (stdout, stderr []byte, err error) {
+func (client *Client) RunSnapctl(options hookstate.SnapCtlRequest) (stdout, stderr []byte, err error) {
```
Comment (niemeyer) [inline client/snapctl.go]: Options was right on this end, per other similar methods. If the point is reusing the same type, it's okay to keep it as options everywhere.

This should also be received as a pointer, as the other option values in client.

Comment (niemeyer) [inline client/snapctl.go]: Also note that, as we agreed before, when reusing such types we should have them inside client rather than inside the other internal package, so it's obvious when we're adding code that makes no sense for the client end.

**Comment by niemeyer** [comment]
LGTM with these last few points sorted.

File: daemon/api.go (hunk 13 above)
Comment (kyrofa) [inline daemon/api.go]: Done.

File: overlord/hookstate/context.go (hunk 14 above)
Comment (kyrofa) [inline overlord/hookstate/context.go]: Done.

File: overlord/hookstate/context.go (hunk 15 above)
Comment (kyrofa) [inline overlord/hookstate/context.go]: Done.

File: overlord/hookstate/context_test.go (hunk 16 above)
Comment (kyrofa) [inline overlord/hookstate/context_test.go]: Done.

File: overlord/hookstate/ctlcmd/ctlcmd.go (hunk 17 above)
Comment (kyrofa) [inline overlord/hookstate/ctlcmd/ctlcmd.go]: Good call, done.

File: overlord/hookstate/ctlcmd/ctlcmd.go (hunk 18 above)
Comment (kyrofa) [inline overlord/hookstate/ctlcmd/ctlcmd.go]: Done.

File: client/snapctl.go (hunk 20 above)
Comment (kyrofa) [inline client/snapctl.go]: Moved this back to `client.SnapCtlOptions`, and used that throughout for consistency.

File: tests/main/snapctl/task.yaml [hunk 21]
```diff
@@ -55,12 +55,20 @@ restore: |
...
+    # This context is enough to hit the API, but the hook will fail since this
+    # context is obviously invalid. That failure still means we're correctly
+    # hitting the API, though, which is all we care about here.
+    export SNAP_CONTEXT="foo"
+    if output="$(snap run --hook=apply-config snapctl-hooks)"; then
+        echo "Expected the hook to fail"
         exit 1
     fi
 
+    if [[ ! "$output" =~ .*"no context for ID: \"foo\"".* ]]; then
```
Comment (niemeyer) [inline tests/main/snapctl/task.yaml]: Spread test still broken here. I'd suggest the more usual:

`echo "$output" | grep -q 'no context for ID: "foo"'`

No need for ifs or the extra message here. Note that English sentence below is a direct translation of the code, which in this specific case is more clear.

Comment (niemeyer) [inline tests/main/snapctl/task.yaml]: Or alternatively, using the bash expression syntax, but simply:

`[[ "$output" =~ 'no context for ID: "foo"' ]]`

I've just confirmed that it traces properly as well, and actually better than echo+grep.

Comment (niemeyer) [inline tests/main/snapctl/task.yaml]: Ah, and the problem is actually on the former line:
```
error: cannot run snapctl: cannot run snapctl: no context for ID: "foo"
+ output='Running snapctl. It should be in my PATH and not cause denials.'
+ [[ ! Running snapctl. It should be in my PATH and not cause denials. =~ .*no context for ID: "foo".* ]]
```

Need something like `output="$(snap ... 2>&1 >/dev/null)"` to capture only stderr.

Comment (kyrofa) [inline tests/main/snapctl/task.yaml]: Oops, good catch, thanks!

File: overlord/hookstate/hooktest/handler.go [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+	DoneError  bool
+
+	ErrorCalled bool
+	ErrorError  bool
+	Err         error
+}
+
+// NewMockHandler returns a new MockHandler.
+func NewMockHandler() *MockHandler {
+	return &MockHandler{
```
Comment (mvo5) [inline overlord/hookstate/hooktest/handler.go]: Pardon my ignorance here, but given goes default policy of zeroing values, why explicatively init them here? Or am I missing something?

**Comment by mvo5** [comment]
Looks good, just one question inline

File: overlord/hookstate/hooktest/handler.go (hunk 22 above)
Comment (kyrofa) [inline overlord/hookstate/hooktest/handler.go]: No, you're not missing anything, I'm being dumb. Fixed now, thanks!
